github-auth-provider/pkg/config/config.go

```diff
@@ -0,0 +1,176 @@
+	ClientID                 string  `env:"OBOT_GITHUB_AUTH_PROVIDER_CLIENT_ID"`
+	ClientSecret             string  `env:"OBOT_GITHUB_AUTH_PROVIDER_CLIENT_SECRET"`
+	ObotServerURL            string  `env:"OBOT_SERVER_URL"`
+	PostgresConnectionDSN    string  `env:"OBOT_AUTH_PROVIDER_POSTGRES_CONNECTION_DSN" optional:"true"`
+	AuthCookieSecret         string  `usage:"Secret used to encrypt cookie" env:"OBOT_AUTH_PROVIDER_COOKIE_SECRET"`
+	AuthEmailDomains         string  `usage:"Email domains allowed for authentication" default:"*" env:"OBOT_AUTH_PROVIDER_EMAIL_DOMAINS"`
+	AuthTokenRefreshDuration string  `usage:"Duration to refresh auth token after" optional:"true" default:"1h" env:"OBOT_AUTH_PROVIDER_TOKEN_REFRESH_DURATION"`
+	GitHubOrg                *string `usage:"restrict logins to members of this GitHub organization" optional:"true" env:"OBOT_GITHUB_AUTH_PROVIDER_ORG"`
+	GitHubAllowUsers         *string `usage:"users allowed to log in, even if they do not belong to the specified org and team or collaborators" optional:"true" env:"OBOT_GITHUB_AUTH_PROVIDER_ALLOW_USERS"`
+}
+
+// Options is the public struct that holds validated and processed configuration values
+type Options struct {
+	options
+	AuthEmailDomains         []string
+	AuthTokenRefreshDuration time.Duration
+	GitHubOrg                string
+	GitHubAllowUsers         []string
+}
+
+var (
+	gitHubLoginRegex = regexp.MustCompile(`^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$`)
+	emailDomainRegex = regexp.MustCompile(`^[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?)*$`)
+)
+
+// LoadEnv loads environment variables, validates them, and returns the completed configuration
+func LoadEnv() (*Options, error) {
+	completed, err := loadEnv()
+	if err != nil {
+		return nil, env.ValidationError{Err: err}
+	}
+
+	return completed, nil
+}
+
+func loadEnv() (*Options, error) {
+	var opts options
+	if err := env.LoadEnvForStruct(&opts); err != nil {
+		return nil, fmt.Errorf("failed to load environment variables: %w", err)
+	}
+
+	return complete(opts)
+}
+
+func complete(o options) (*Options, error) {
+	var (
+		validationErrors env.FieldValidationErrors
+		completedOptions = Options{
+			options: o,
+		}
+	)
+
+	if o.AuthEmailDomains != "" {
+		// TODO(njhale): Add validation for email domains
+		var (
+			emailDomains = strings.Split(o.AuthEmailDomains, ",")
+			errorMsg     string
+		)
+		for i := range emailDomains {
+			switch domain := strings.TrimSpace(emailDomains[i]); {
+			case domain == "":
+				errorMsg = "cannot contain empty email domains"
+			case domain == "*" && len(emailDomains) > 1:
+				errorMsg = "cannot specify multiple email domains when * is provided"
+			case domain != "*" && !emailDomainRegex.MatchString(domain):
+				errorMsg = fmt.Sprintf("'%s' is not a valid email domain", domain)
+			default:
+				emailDomains[i] = domain
+				continue
+			}
+
+			// Stop after the first email domain validation error
+			break
+		}
```

Comment:
Pretty sure this configuration is common to all auth providers. It has me wondering if part of this validation /completion should live in `env.LoadEnv`; e.g. you build a validator/completor to pass into `env.LoadEnv`.